# here are some links to articles we should read

*  *Detecting and Estimating Contamination of Human DNA Samples in Sequencing and Array-Based Genotype Data*
This is an article in AJHG from Boehnke's group.  http://www.ncbi.nlm.nih.gov/pmc/articles/PMC3487130/  It is geared toward human stuff for sure and combines array and sequencing methods.


*  *Here is a google scholar "related-articles" search for the above*  http://scholar.google.com/scholar?bav=on.2,or.r_cp.r_qf.&bvm=bv.67720277,d.cGU,pv.xjs.s.en_US.u4s5c7_Kv3Q.O&biw=1381&bih=738&dpr=1&um=1&ie=UTF-8&lr=&q=related:Vk_ZjMpPmK3cIM:scholar.google.com/


* *ContEst: estimating cross-contamination of humnan samples next-generation sequencing data* 
http://bioinformatics.oxfordjournals.org/content/27/18/2601  As the title says this is geared towards human sequencing. Bayesian approah to calculate the posterior probability of the contamination level.


* *A Novel Method for Detecting Contaminated Samples Based on Illumina Sequencing Data*
http://www.ijbbb.org/papers/322-E0014.pdf  Model for detecting contamination using mappability of sites and SNP sites on the X and Y chromosomes.


* *Source-Sink Estimates of Genetic Introgression Show Influence of Hatchery Strays on Wild Chum Salmon Populations in Prince William Sound, Alaska*
http://www.plosone.org/article/info%3Adoi%2F10.1371%2Fjournal.pone.0081916#pone-0081916-t002  Study from the Alaska Department of Fish and Game. They used an MCMC model and microsatellites to identify contaminated samples.  The two methods were not very consistent. The methods were used for quality control, so they did not run simulations or test to determine the accuracy of either method.

